### ISSUE TITLE
scans not sorted in retention time dimension (Waters data)

### ISSUE THREAD

**Issue by niekdejonge** [mzmine/mzmine3#1277]
### Basic information

* My operating system and version: Windows 
* My MZmine version: MZMine 3.4.16

### What happened

When building a chromatogram from MZML files I get the error:
Scans not sorted in retention time dimension! Cannot create chromatogram. 
The files came from a Waters instrument. 

Suggested change:
- Add a module to sort the spectra on retention time. 
- Instead of the current warning, give a warning that suggests to use the described module above. 

To solve the issue locally I wrote a quick python script to replace the scan with the spectrum index. This solved the issue. 

```python
import re

with open("../../data/urine_fractions_london/mzML/FractionProfiling_RPOS_ToF10_Fraction001_DDA.mzML", "r") as file:
    content = file.read()
    content_new = re.sub('<spectrum index="(\d+)" (.+ scan=)(\d+)"', r'<spectrum index="\1" \2\1"', content, count=0)

with open("../../data/urine_fractions_london/mzML/FractionProfiling_RPOS_ToF10_Fraction001_DDA_corrected.mzML", "w") as file:
    file.write(content_new)
```

**Comment by Bevoniezh** [comment]
Hi @niekdejonge, I also have this issue. I created a homemade python code to fix it. It both correct the scan numbers and also associate the ms level to the correct func file from the .raw folder if it has failed uring the conversion (it should not, but in some old version of MSconvert it could happen).
It takes 3-20 seconds to do it depending on the size of your .mzml

`    
import re

def adjust(self, File): # File is your converted file name from MSconvert
    with open(mzml, 'r') as f: # Open a new file for writing        
        with open(mzml[:-5]+'_adjusted.mzML', 'w') as output_file:  # this create your file with adjusted scan numbers
            count_line = 0
            ref_line = -1 # This object will help us not to change the lines with "ms level"
            for line in f:
               # Use a regular expression to check if the line matches the pattern "spectrum index=" and contains "scan="
                if re.search(r'spectrum index="(\d+)"', line) and "scan=" in line:
                    spectrum_index = int(re.search(r'spectrum index="(\d+)"', line).group(1)) # Extract the number following "spectrum index="
                    new_scan = spectrum_index + 1 # Calculate the new value of "scan=" as spectrum_index + 1
                    new_line = re.sub(r'scan=\d+', f'scan={new_scan}', line) # Replace the current value of "scan=" with the new value
                    # after, we adjust the ms level
                    ms_level = int(re.search(r'id="function=(\d+)', line).group(1)) # Find the true ms level based on the func data file name from raw files
                    ref_line = count_line + 2
                    output_file.write(new_line) # Write the updated line to the output file                        
                else:
                    if count_line == ref_line:
                         new_line = re.sub(r'value="(\d+)"', f'value="{ms_level}"', line)
                         output_file.write(new_line) 
                    else:
                        output_file.write(line) # If the line doesn't match the pattern, write it to the output file unchanged                        
                count_line += 1
'

Tell me if it does not work

**Comment by Bevoniezh** [comment]
Yes it is true, as I told in some conversions there is an issue concerning the ms level (it happened in my group work). When you convert there are only ms level = 1 even for the ms level = 2. It can happen when you use old versions of MSconvert in centralized servers. I agree it overcomplicates the code but is to handle even more possibilities because sometimes the scan number is not the only issue after conversion, and I found it relevant to put my comment in this topic. But I prefer your code concerning the scan number, it is much cleaner.

**Comment by niekdejonge** [comment]
Ah that is good to know, thanks. In that case it is indeed best to make it as robust as possible.

**Comment by GSH-09** [comment]
> It can happen when you use old versions of MSconvert in centralized servers.

Just double checking, but where is the correct place to download the latest version?

I am running into this problem as well with Full Scan + DDA data. As suggested on another thread, I tried Raw Data Filter > Crop Filter > MS1 only, but the error still shows up.

**Comment by kirankumarp20** [comment]
Hello @Bevoniezh tried with this python able to download both adjusted and corrected mzml file but unable to import into mzmine application showing this error (check png).

![image](https://github.com/mzmine/mzmine3/assets/75622361/e75789a4-7387-4444-9104-595c0e711009)
 
can you help with this. I am unable to complete and can't compete with deadline and newer to all this.

![Capture10](https://github.com/mzmine/mzmine3/assets/75622361/63e7b1c1-5fa9-4e04-8f29-ee87537d2c8c)
 
this issues were also coming same time when using MSconvert gui .

**Comment by mhagar** [comment]
# How to fix:
I've encountered and solved this issue, and was directed by a colleague to share the fix here for the search-engines, etc.

Improperly converted Waters datasets will contain lockmass scans interleaved with the actual stuff you're interested in. This will cause errors.
![image](https://github.com/user-attachments/assets/c1d6e399-0a30-431a-85c9-21e2fc3ac122)

To convert Waters `.RAW` files into well-formed `.mzmL` files, you need to use these settings on MSConvert:
![image](https://github.com/user-attachments/assets/52a8092e-d956-419e-9fe1-022cf7595e60)

To figure out what Scan events to keep, you can check the `.RAW` files:
![image](https://github.com/user-attachments/assets/2fa3e722-3cc9-47ad-84ba-bdf083256708)

Note that you can save these settings in the MSConvert GUI, for future convenience (bottom left corner). Allegedly, Waters has an official [tool](https://microapps.on-demand.waters.com/home/showmarkdown/data-as-a-product) for converting their data into `.mzmL`, but I haven't been able to download it. If you can't download it either, and using MSConvert as I've described is overly inconvenient for you, please join me in harassing them at WatersMicroApps@waters.com

**Comment by SteffenHeu** [comment]
Hi @mhagar, thanks for looking into this and providing a solution. We have worked on using MSConvert for import of vendor files into mzmine, which is now available in the 4.3 release. We try to auto detect the location of MSConvert, otherwise you need to set it in the preferences. 
Similarly, we have optimised the import of native Waters data with regard to the parameters we use for the conversion. 
Please try the new import and if it leads to satisfactory results (it did in our testing).

**Comment by mhagar** [comment]
> Hi @mhagar, thanks for looking into this and providing a solution. We have worked on using MSConvert for import of vendor files into mzmine, which is now available in the 4.3 release. We try to auto detect the location of MSConvert, otherwise you need to set it in the preferences. Similarly, we have optimised the import of native Waters data with regard to the parameters we use for the conversion. Please try the new import and if it leads to satisfactory results (it did in our testing).


Hi @SteffenHeu, perfect timing! Auto-controlling `msconvert` is a fantastic idea. I've given importing Waters data with 4.3 a brief test, here's how it worked:

- Using the file-browser pop-up upon clicking 'Select Files', I couldn't select the Waters `.RAW` folders (even while setting the browser to specifically show that type). The browser would enter the folder rather than select it.
	- This occured in both Linux and Windows
	- I got around it by just typing in the directory into the main MS import dialogue

- On Linux, I was not able to direct MZMine to the `msconvert` executable. The file-browser simply does not detect it, and there's no way to manually input the path. The situation is more clear when trying to set the path from the Preferences dialogue (Ctrl P), as the file-browser explicitly states that it's looking for Microsoft Windows executables. I suspect it would not be too difficult to someday implement an option for using the `msconvert` Docker image for Mac/Linux users.

- On Windows, `msconvert` _was detected seamlessly_ and all the steps in the standard `.mzmL` -> processed feature-list pipeline worked with no problems.
	- However, I noticed a minor inconvenience: the first page of the `mzwizard` doesn't allow for selection of `.mzmL` files that are already imported, and I had to re-import the Waters `.RAW` folders.

All in all, I think this is a great feature and will streamline things quite a bit

**Comment by SteffenHeu** [comment]
Thanks for testing!

>     * Using the file-browser pop-up upon clicking 'Select Files', I couldn't select the Waters `.RAW` folders (even while setting the browser to specifically show that type). The browser would enter the folder rather than select it.

Since the popup is for files, the selection of folders is not possible. (The same goes for Bruker data). We recommend to drag and drop the files. (Either into the main window or into the text field of the dialogs)
 
>     * On Linux, I was not able to direct MZMine to the `msconvert` executable. The file-browser simply does not detect it, and there's no way to manually input the path. The situation is more clear when trying to set the path from the Preferences dialogue (Ctrl P), as the file-browser explicitly states that it's looking for Microsoft Windows executables. I suspect it would not be too difficult to someday implement an option for using the `msconvert` Docker image for Mac/Linux users.

Good catch, i will update the dialog. Using a docker image is a good idea, but I suspect if the user is able to install and run docker, he can probably convert the files him/herself :D.

>       * However, I noticed a minor inconvenience: the first page of the `mzwizard` doesn't allow for selection of `.mzmL` files that are already imported, and I had to re-import the Waters `.RAW` folders.

I'm not totally sure what you mean here. Can you provide an example?
